### ISSUE TITLE
zstash reads stored index.db instead of retrieving the new one specified by --hpss

### ISSUE THREAD

**Issue by wenshanw** [E3SM-Project/zstash#100]
Hi zstash team,

I tried to list files from two different HPSS locations.
It seems zstash uses the first index.db even though I specified a new location using --hpss.
The command I use is `zstash ls --hpss=[location 1/2]`.

How can I avoid this problem?
I found it a little bit dangerous for new users.

Thank you!

**Comment by chengzhuzhang** [comment]
Hi @wenshanw, For two simulations, I usually create separate directories for index.db and to extract files from each simulations. You may want to try issue `zstash` commands at different directories.

**Comment by wenshanw** [comment]
Thank you @chengzhuzhang!
I will keep it in mind in the future.
It will be great if zstash could retrieve a new index.db once the hpss location is different or pop up some messages for the new users.

**Comment by tangq** [comment]
@chengzhuzhang , will deleting the existing index.db file before extracting the one on HPSS cause any issue? If not, it seems a reasonable way to handle it.

Otherwise, an alternative is printing out a warning message indicating that the old index.db is used and if the user wants to extract the HPSS copy, one should change the directory or delete the .db file manually and rerun the command.

**Comment by maltrud** [comment]
i totally agree with @tangq on this point.  i think it is a bad idea to delete the local version without warning--that would have hosed me more than once during v1.

**Comment by chengzhuzhang** [comment]
Thank you @tangq and @maltrud. I agree that we need a proper warning in place when users try to overwrite with a new index.bd.

**Comment by forsyth2** [comment]
Hi @wenshanw thanks for reporting the issue. I want to make sure I'm interpreting your problem correctly. Do you (or @chengzhuzhang) have exact steps to reproduce the error?

I created an example directory `zstash_demo` and ran the following commands:
```
zstash create --hpss=zstash_archive zstash_demo # zstash_demo/zstash/ contains index.db
zstash ls --hpss=zstash_archive # produces directory zstash, which contains index.db
# <Create a new file in the zstash_demo directory>
zstash create --hpss=zstash_archive2 zstash_demo 
zstash ls --hpss=zstash_archive2 # Does NOT include the file we just added
rm -rf zstash
zstash ls --hpss=zstash_archive2 # Now, DOES include the file we just added
```
However, I'm not sure those are the steps you have in mind. If these are the steps you have in mind, you're saying we should make it such that the first `zstash ls --hpss=zstash_archive2` would have included the new file, correct?

**Comment by wenshanw** [comment]
Hi @forsyth2, thank you very much for creating this example to reproduce my issue!
Yes, these are basically the commands I have in mind.

The problem for me is that I have no idea what files are included in the HPSS archives.
At my home directory, I `zstash ls` archive1 and archive2 but neither seemed to have the files I was looking for.
In fact, archive2 has my files. I didn't realize I `zstash` the same index.db with different `--hpss` locations. 

Yes, I would expect `zstash ls` to list files in the path specified by `--hpss` no matter at which directory I execute `zstash`.

**Comment by forsyth2** [comment]
@wenshanw Thanks, I've marked this issue to be addressed before the next release of zstash. 

It looks like `zstash` is using the `index.db` in the already extracted directory. In fact, this issue appears to be a duplicate of #68 created by @TonyB9000.

In the meantime, you can always check the contents of an HPSS archive by logging into `hsi`, and running `ls <archive-name>`.

**Comment by wenshanw** [comment]
Hi @forsyth2, I think my problem is a little different from #68. Since Tony did not specify `--hpss`, I would expect `zstash` to read the existing `index.db`. But when I use `--hpss`, I would expect `zstash` to retrieve a new `index.db`. 
Maybe these are the two sides of a coin.

By using `ls <archive-name>` after logging into `hsi`, I can only see tar files, right? not as helpful as what I can see by using `zstash ls` :)

**Comment by forsyth2** [comment]
@wenshanw Oh yes, you're right on both statements. I do imagine that #68 and this issue can be addressed by similar fixes, if not the same one.

**Comment by TonyB9000** [comment]
Side note:  The term “archive” is overloaded, in part due to me, perhaps.

There is the (hsi) HPSS “the archive”,
Upon extraction, there is a locally-created “zstash” directory, with tarfiles and index.db, also referred to as “the archive”
Lastly, files extracted (according to their tarfile-paths) are also sometimes referred to as “the archive”.

With both “create” and “extract”, with then without --hpss, all of these terms become conflated for me.

Need flowchart-graphics to keep it straight for me.

OR, refer exclusively to one of
                hpss-archive
                local-archive
                the_files (source under “create”, dest under “extract”)

**Comment by forsyth2** [comment]
Yes, we definitely use the term `archive` quite a lot. We tried to clear that up in the Terminology section of the tutorial (https://e3sm-project.github.io/zstash/docs/html/tutorial.html#terminology):

`zstash` archive: A set of files {`index.db`, `nnnnnn.tar` where nnnnnn=000000, 000001, …}
HPSS: A long-term tape storage system.
HPSS archive: The `zstash` archive on HPSS.
Local archive (or cache): The `zstash` archive on our local file system. The default name is `zstash`.
Source directory: the directory whose contents we are archiving.

**Comment by TonyB9000** [comment]
Source directory: the directory whose contents we are archiving.

We should probably add

                Dest(ination) directory where contents are being extracted

Neither Source directory nor Dest directory should be referred to as (an,the) archive.

____tony____

**Comment by golaz** [comment]
@wenshanw, thanks for reporting this. I did not think about the use-case of a user exploring the content of multiple zstash simulation archives to find files they are looking for. I was working under the assumption of a user wanting to find specific files within a given simulation.

Internally, zstash will only retrieve `index.db` from hsi if it is not already present in the local cache (zstash/ sub-directory) which is what led to the confusion above. 

Currently, there is no consistency check. We will consider adding a consistency check between the local and remote copy of the database in a future version of zstash, if we can find a way to do so without adding too much complexity.

In the mean time, if you change --hpss between calls to zstash, either delete the zstash cache sub-directory or execute zstash from another sub-directory.

**Comment by golaz** [comment]
Won't change the code for now, but add some note in the documentation.

**Comment by forsyth2** [comment]
Closed by #113.
